### ISSUE TITLE
Fake quantization produces diverse outputs under CPU and CUDA backends.

### ISSUE THREAD

**Issue by zeakey** [pytorch/pytorch#58130]
## 🐛 Bug

The `fake quantization` produces different outputs between  GPU and CPU.  In my understanding,
the behaviour of `fake quantization` should be identical no matter the model is on  CPU or GPU.

## Reproduce

Code snippet to reproduce the bug:
```
import torch
import torch.nn as nn
import torchvision

torch.manual_seed(0)

model = torchvision.models.quantization.resnet.resnet18().cuda()
model.fuse_model()
model.qconfig = torch.quantization.get_default_qat_qconfig("fbgemm")
torch.quantization.prepare_qat(model, inplace=True)

x = torch.rand(2, 3, 224, 224)
y0 = model(x.cuda())
y1 = model.to(device=torch.device("cpu"))(x)
print(y0)
print(y1)
```

outputs:
```
tensor([[ 0.2143,  0.2571, -0.3428,  ..., -0.0857, -0.0429,  0.9214],
        [ 0.2786,  0.2143, -0.1929,  ...,  0.0643,  0.0000,  1.0714]],
       device='cuda:0', grad_fn=<FakeQuantizePerTensorAffineCachemaskBackward>)


tensor([[ 0.2143,  0.2357, -0.3215,  ..., -0.0214, -0.0429,  0.9644],
        [ 0.2786,  0.2357, -0.2143,  ...,  0.0429,  0.0214,  1.0929]],
       grad_fn=<FakeQuantizePerTensorAffineCachemaskBackward>)
```

When commenting `torch.quantization.prepare_qat(model, inplace=True)`, the outputs are matched between CPU and GPU.

## Expected behavior

The output should be exactly identical.

## Environment
```
PyTorch version: 1.8.1+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Tencent tlinux 2.2 (Final) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-5)
Clang version: Could not collect
CMake version: version 3.19.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GRID V100DX-8C
Nvidia driver version: 430.46
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.2
[pip3] torch==1.8.1+cu101
[pip3] torchvision==0.9.0
[conda] numpy                     1.20.2                   pypi_0    pypi
[conda] torch                     1.8.1+cu101              pypi_0    pypi
[conda] torchvision               0.9.0                    pypi_0    pypi
```

___

## UPDATE

Using `get_default_qconfig` instead of `get_default_qat_qconfig` resolves the problem:
```
import torch
import torch.nn as nn
import torchvision

torch.manual_seed(0)

model = torchvision.models.quantization.resnet.resnet18().cuda()
model.fuse_model()
model.qconfig = torch.quantization.get_default_qconfig("fbgemm")
torch.quantization.prepare_qat(model, inplace=True)

x = torch.rand(2, 3, 224, 224)
y0 = model(x.cuda())
y1 = model.to(device=torch.device("cpu"))(x)
print(y0)
print(y1)
```
Outputs:
```
tensor([[ 0.2437,  0.2732, -0.2258,  ...,  0.0304, -0.0577,  1.0358],
        [ 0.1577,  0.2105, -0.3000,  ..., -0.0177,  0.0316,  0.9406]],
       device='cuda:0', grad_fn=<AddmmBackward>)
tensor([[ 0.2437,  0.2732, -0.2258,  ...,  0.0304, -0.0577,  1.0358],
        [ 0.1577,  0.2105, -0.3000,  ..., -0.0177,  0.0316,  0.9406]],
       grad_fn=<AddmmBackward>)
```

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo

**Comment by zeakey** [comment]
The mobilenets presents much larger difference in outputs compared to resnets.

```
import torch
import torch.nn as nn
import torchvision

torch.manual_seed(0)

model = torchvision.models.quantization.mobilenet_v2().cuda()
model.fuse_model()
model.qconfig = torch.quantization.get_default_qat_qconfig("fbgemm")
torch.quantization.prepare_qat(model, inplace=True)

x = torch.rand(2, 3, 224, 224)
y0 = model(x.cuda())
y1 = model.to(device=torch.device("cpu"))(x)

print(torch.nn.functional.cosine_similarity(y0, y1.cuda()))
```
outputs `tensor([0.7789, 0.7733], device='cuda:0', grad_fn=<DivBackward0>)`.
When the model is `torchvision.models.quantization.resnet.resnet18()`, the cosine similarity is 0.98.

Since I'm working on face recognition where the cosine similairty between samples are critically sensitive, this difference between  CPU and GPU is non-neglectable.


My full pipeline is as bellow:
quantization-aware training (fake quantization) on CUDA ->  transport model to CPU -> convert to real quantization with `torch.quantization.convert` -> export model to onnx.

The different behaviours between CPU and GPU interupts my pipeline.

**Comment by HDCharles** [comment]
As I understand, the scale and zero point are changing over time. Try the following:

```
import torch
import torch.nn as nn
import torchvision

torch.manual_seed(0)

model = torchvision.models.quantization.mobilenet_v2().cuda()
model.fuse_model()
model.qconfig = torch.quantization.get_default_qat_qconfig("fbgemm")
torch.quantization.prepare_qat(model, inplace=True)

x = torch.rand(2, 3, 224, 224)
y0 = model(x.cuda())
y1 = model(x.cuda())
print(torch.nn.functional.cosine_similarity(y0, y1))
```

I get 

```tensor([0.7744, 0.7989], device='cuda:0', grad_fn=<DivBackward0>)```

**Comment by zeakey** [comment]
@HDCharles the scale and zero-point are a problem.
By disabling the observer,
```
import torch
import torch.nn as nn
import torchvision

torch.manual_seed(0)

model = torchvision.models.quantization.resnet.resnet18().cuda()
model.fuse_model()
model.qconfig = torch.quantization.get_default_qat_qconfig("fbgemm")
torch.quantization.prepare_qat(model, inplace=True)

x = torch.rand(2, 3, 224, 224)
model(x.cuda())

model.apply(torch.quantization.disable_observer)

y0 = model(x.cuda())
y1 = model.to(device=torch.device("cpu"))(x)

print(torch.nn.functional.cosine_similarity(y0, y1.to(y0)))
```
I got `tensor([0.9988, 0.9988], device='cuda:0', grad_fn=<DivBackward0>)`.

**Comment by HDCharles** [comment]
To clarify a bit further,

there is no guarantee between CUDA and CPU bitwise equivalence, though they should be extremely close. 

The above code does have a separate issue in that the batchnorms are still updating. You can get around this with copy.deepcopy

```
import torch
import torchvision
import copy

torch.manual_seed(0)

model = torchvision.models.quantization.resnet.resnet18().cuda()
model.train()
model.fuse_model()
model.qconfig = torch.quantization.get_default_qat_qconfig("fbgemm")
torch.quantization.prepare_qat(model, inplace=True)

x = torch.rand(2, 3, 224, 224)
model(x.cuda())
model1 = copy.deepcopy(model)

y0 = model(x.cuda())
y1 = model1.to(device=torch.device("cpu"))(x)

print(torch.nn.functional.cosine_similarity(y0, y1.to(y0)))


model2 = torchvision.models.quantization.resnet.resnet18()
model2.train()
model2.fuse_model()
model2.qconfig = torch.quantization.get_default_qat_qconfig("fbgemm")
torch.quantization.prepare_qat(model2, inplace=True)

x = torch.rand(2, 3, 224, 224)
model2(x)
model3 = copy.deepcopy(model2)

y2 = model2(x)
y3 = model3(x)

print(torch.nn.functional.cosine_similarity(y2, y3))
```

The result of which is: 

```
tensor([0.9995, 0.9995], device='cuda:0', grad_fn=<DivBackward0>)
tensor([1., 1.], grad_fn=<DivBackward0>)
```

**Comment by zeakey** [comment]
@HDCharles thx for clarifying. things seem to be clear now.
